### ISSUE TITLE
A complex data type for 'referenced' identifiers?

### ISSUE THREAD

**Issue by mbrush** [ga4gh/va-spec#60]
Consider the possibility of a complex identifier data type that includes fields for the identifier itself (code), and basic/generic information to help a user understand what the identified/referenced entity is (e.g. 'entityType' and 'label') . 

This is particularly relevant for us if we give ACM-based names to the attributes used to structure statement semantics (i.e. 'subject', and  'descriptor', rather than 'variant' and 'disease') - as these names which do not tip the user off as to what type of thing the attribute holds. In the case that the identifier in a subject slot is opaque (e.g. clingen:12345), and the entity's representation is not inlined - a user may not know from the data itself what type of thing clingen:12345 is.  This potential issue could be addressed if the identifier is wrapped in a complex data type that includes the ability to provide a type and label.

**Comment by mbaudis** [comment]
@mbrush +1; this is the standard we had agreed on in various GA4GH schema parts. Basically using `OntologyClass` [objects / structure](https://schemablocks.org/schemas/sb-phenopackets/OntologyClass.html) for identifiers (even if not strictly ontologies); the label should be optional:

```
{
   "id" : "DUO:0000004",
   "label" : "no restriction"
},
{
   "id" : "cellosaurus:CVCL_2253",
   "label" : "cell line C-4-I"
},
{
    "id" : "PMID:28966033"
}
```

**Comment by rrfreimuth** [comment]
HL7 has used a complex type for Identifier for many years (see [here](https://www.hl7.org/fhir/datatypes.html#Identifier)).  IMO, the core part that we should support at a minimum is:

```
type     CodeableConcept   Coded type describing the identifier
system   uri               Namespace for the identifier
value    string            Value of the identifier
```

Note that CodeableConcept is also a complex type:

```
coding   Coding   Code defined by a terminology system
text     string   Plain text representation of the concept (used when a Coding is not available)
```

Coding is also a complex type:

```
system    uri      Identity of the terminology system
version   string   Version of the system - if relevant
code      code     Symbol in syntax defined by the system (string + syntax rules)
display   string   Representation defined by the system
```

Example (illustrating two standardized codes from different systems that are mapped to a codeableconcept from my favorite local code system):

```
<Identifier>
  <system value="http://my.favorite.system" />
  <value value="123456" />
  <type>
    <coding>
      <system value="http://hl7.org/fhir/sid/icd-10" />
      <code value="R51" />
    </coding>
    <coding>
      <system value="http://snomed.info/sct" />
      <code value="25064002" />
      <display value="Headache" />
    </coding>
    <text value="general headache" />
  </type>
</Identifier>
```

I realize this is a much more complex structure than most GA4GH project might need, but it isn't too different from the `Resource` and `OntologyClass` structures in Phenopackets.

**Comment by mbaudis** [comment]
@rrfreimuth But the basic code's 3 components correspond ~ to the 2 in `OntologyClass`; `system:value` are represented through the `id` there (recommended format: CURIE). Now, in contrast to the "everything is resolved through the CURIE, `CodeableConcept` has its own built-in structure; but in principle this maps.

**Comment by mbrush** [comment]
Just to clarify, the proposal here is for representing identifiers for referenced instances in the data when they are captured 'by reference' instead of inlined.  I think this is analogous to how the FHIR '[Reference](http://hl7.org/fhir/references.html#Reference)' data type is used - e.g. to capture a related Encounter in the Observation resource [here](http://hl7.org/fhir/observation-definitions.html#Observation.encounter).

This is separate issue from how we propose to capture terms from ontologies/terminologies as coded values for attributes - which is where FHIR CodableConcepts and Codings come into play.  The proposed VA ['coding' data type](https://docs.google.com/document/d/1yL65lGqx_ZFIBOkUOy2Na79qRSIgS24D0Rvac6-wUHQ/edit#heading=h.xay81vy07llj) is used for this, and is largely modeled after a FHIR 'coding'. 

What I am talking about here is a way to capture 'references' to other instances in a data set as a structured object using more than the identifier for the referenced object. i.e. bundling this with its type and perhaps even its label.  So that users will have a sense for what it is when its representation is not inlined.  I think a better name for this proposed complex data type would be to follow FHIR and call it a 'Reference' - updating the issue title accordingly.

(As an aside the full proposal for initial data types in the VA Spec is [here ](https://docs.google.com/document/d/1yL65lGqx_ZFIBOkUOy2Na79qRSIgS24D0Rvac6-wUHQ/edit) - and will be discussed on the VA call today in about an hour.   The 'identifier' data type here is a simple data type for representing the string used to capture an instance identifier in the data.)

**Comment by mbaudis** [comment]
@mbrush Yes, I got that. But IMO that is the case here for "Id =~ URI/CURIE" + some label / description.
